machine/v1alpha1/provider_types.go

```diff
@@ -0,0 +1,207 @@
+	// Name of the container. Cannot be updated.
+	Name string `json:"name"`
+
+	// Container Image Name
+	// +optional
+	Image *ImageMeta `json:"image,omitempty"`
+
+	// Args represents extra provider specific flags that are not encoded as fields in this API.
+	// Explicit controller manager properties defined in the `Provider.ManagerSpec`
+	// will have higher precedence than those defined in `ContainerSpec.Args`.
+	// For example, `ManagerSpec.SyncPeriod` will be used instead of the
+	// container arg `--sync-period` if both are defined.
+	// The same holds for `ManagerSpec.FeatureGates` and `--feature-gates`.
+	// +optional
+	Args map[string]string `json:"args,omitempty"`
+
+	// List of environment variables to set in the container.
+	// +optional
+	Env []corev1.EnvVar `json:"env,omitempty"`
+
+	// Compute resources required by this container.
+	// +optional
+	Resources *corev1.ResourceRequirements `json:"resources,omitempty"`
+}
+
+// ImageMeta allows to customize the image used.
+type ImageMeta struct {
+	// Repository sets the container registry to pull images from.
+	// +optional
+	Repository *string `json:"repository,omitempty"`
+
+	// Name allows to specify a name for the image.
+	// +optional
+	Name *string `json:"name,omitempty"`
+
+	// Tag allows to specify a tag for the image.
+	// +optional
+	Tag *string `json:"tag,omitempty"`
+}
+
+// FetchConfiguration determines the way to fetch the components and metadata for the provider.
+type FetchConfiguration struct {
+	// URL to be used for fetching the provider’s components and metadata from a remote Github repository.
+	// For example, https://github.com/{owner}/{repository}/releases
+	// The version of the release will be `ProviderSpec.Version` if defined
+	// otherwise the `latest` version will be computed and used.
+	// +optional
+	URL *string `json:"url,omitempty"`
+
+	// Selector to be used for fetching provider’s components and metadata from
+	// ConfigMaps stored inside the cluster. Each ConfigMap is expected to contain
+	// components and metadata for a specific version only.
+	// Note: the name of the ConfigMap should be set to the version or to override this
+	// add a label like the following: provider.cluster.x-k8s.io/version=v1.4.3
+	// +optional
+	Selector *metav1.LabelSelector `json:"selector,omitempty"`
+}
+
+// ProProviderStatus defines the observed state of the Provider.
+type ProviderStatus struct {
+	// Contract will contain the core provider contract that the provider is
+	// abiding by, like e.g. v1alpha4.
+	// +optional
+	Contract *string `json:"contract,omitempty"`
+
+	// Conditions define the current service state of the provider.
+	// +optional
+	Conditions clusterv1.Conditions `json:"conditions,omitempty"`
+
+	// ObservedGeneration is the latest generation observed by the controller.
+	// +optional
+	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+
+	// InstalledVersion is the version of the provider that is installed.
```

Comment:
What version does this show during an upgrade if part of the system is v1 and another part is v2?

Reply:
It will show the latest successfully installed version.